### PR TITLE
Part II, Lesson 5

### LINKED ISSUE #16: Part II, Lesson 5
This lesson will introduce object-to-BG interactions. The ball must start colliding against the bricks in the background as well, bouncing and eliminating them. Like for OBJs, VRAM shall be accessed directly to perform collision checks.

This can be made buggy the first time around, e.g. always checking above the ball; the bug can then be demonstrated (collision misbehaving when the ball hits bricks from the side and/or top), and fixed. Again, it's important to show how to debug and iterate on the code.

### PR DESCRIPTION
**Pull request by eievui5**
Closes #16

### CODE REVIEW

File: unbricked/bricks/main.asm [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+	inc hl
+	ld [hl], BLANK_TILE
+CheckForBrickRight:
+	cp a, BRICK_RIGHT
+	ret nz
+	; Break a brick from the right side.
+	ld [hl], BLANK_TILE
+	dec hl
+	ld [hl], BLANK_TILE
+	ret
```
Comment (ISSOtm) [inline unbricked/bricks/main.asm:304]:
```suggestion
	jp z, BreakBrick ; Break a brick from the left side.
	cp a, BRICK_RIGHT
	ret nz
	; Break a brick from the right side.
BreakBrick:
	ld [hl], BLANK_TILE
	dec hl
	ld [hl], BLANK_TILE
	ret
```

File: unbricked/bricks/main.asm [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+	ret z
+	cp a, $06
+	ret z
+	cp a, $07
+	ret
+
+; ANCHOR: check-for-brick
+; Checks if a brick was collided with and breaks it if possible.
+; @param hl: address of tile.
+CheckForBrick:
```
Comment (ISSOtm) [inline unbricked/bricks/main.asm:289]: I don't know if I'm on board with that name: `Check` imo implies that it should *return* whether a brick is there, but have no side effects.
```suggestion
CheckAndHandleBrick:
```

File: unbricked/bricks/main.asm [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+	add a, b
+	ld [_OAMRAM + 5], a
+
+	ld a, [wBallMomentumY]
+	ld b, a
+	ld a, [_OAMRAM + 4]
+	add a, b
+	ld [_OAMRAM + 4], a
+
+; ANCHOR: updated-bounce
```
Comment (ISSOtm) [inline unbricked/bricks/main.asm:118]: Hmm, isn't that a big chunk? Maybe we should introduce diff syntax, instead (like in the pokecrystal wiki).

File: src/part2/bricks.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+We've already used some constants, like `rLCDC` from `hardware.inc`, but we can also create our own for anything we want.
+Let's make three constants at the top of our file, representing the tile IDs of left bricks, right bricks, and blank tiles.
+```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/bricks/main.asm:constants}}
+{{#include ../../unbricked/bricks/main.asm:constants}}
+```
+
+Constants are a kind of *symbol* (which is to say, "a thing with a name").
+Writing a constant's name in an expression is equivalent to writing the number the constant is equal to, so `ld a, BRICK_LEFT` is the same as `ld a, $05`.
+But I think we can all agree that the former is much clearer, right?
+
```
Comment (ISSOtm) [inline src/part2/bricks.md:15]:
```suggestion

## Destroying bricks

```
A title here should help mark the change of topics better.

File: src/part2/bricks.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+Now we'll write a function that checks for and destroys bricks.
+Our bricks are two tiles wide, so when we hit one we'll have to remove the adjacent tile as well.
+If we hit the left side of a brick (represented by `BRICK_LEFT`), we need to remove it and the tile to its right (which should be the right side).
+If we instead hit the right side, we need to remove the left!
```
Comment (ISSOtm) [inline src/part2/bricks.md:19]:
```suggestion
If we instead hit the right side, we need to remove the left!

```

File: src/part2/bricks.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/bricks/main.asm:check-for-brick}}
+{{#include ../../unbricked/bricks/main.asm:check-for-brick}}
+```
+
+Just insert this function into each of your bounce checks now.
+It should go right before you modify the momentum; make sure you don't miss any!
```
Comment (ISSOtm) [inline src/part2/bricks.md:25]:
```suggestion
Make sure you don't miss any!
It should go right **before** the momentum is modified.

```

**Review by ISSOtm — COMMENTED** [review]
Alright, great! Short and sweet (for once :sweat_smile:), clear enough, so we're good.

That said, shouldn't we make another lesson before that one? To explain `IsWallTile` and all of its arithmetic. (@avivace should I open a new issue for that?)

**Comment by avivace** [comment]
> Alright, great! Short and sweet (for once sweat_smile), clear enough, so we're good.
> 
> That said, shouldn't we make another lesson before that one? To explain `IsWallTile` and all of its arithmetic. (@avivace should I open a new issue for that?)

I agree! Please do that

**Review by avivace — APPROVED** [review]

File: unbricked/bricks/main.asm (hunk 3 above)
Comment (eievui5) [reply unbricked/bricks/main.asm:118]: Yeah I think that would be better.

File: unbricked/bricks/main.asm (hunk 1 above)
Comment (eievui5) [reply unbricked/bricks/main.asm:304]: This seems to always break from the right side? Are you missing some `inc hl`s?

**Pull request merged**